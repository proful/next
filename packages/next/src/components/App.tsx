/* eslint-disable @typescript-eslint/no-explicit-any */
import * as React from 'react'
import { observer } from 'mobx-react-lite'
import type { TLAppPropsWithoutApp, TLAppPropsWithApp } from '~types'
import type { TLShape } from '~lib'
import { AppContext, Renderer } from '~components'
import { useApp } from '~hooks'

export type AppProps<S extends TLShape> = TLAppPropsWithoutApp<S> | TLAppPropsWithApp<S>

export function App<S extends TLShape>(props: AppProps<S>): JSX.Element {
  return (
    <AppContext {...props}>
      <InnerApp {...props} />
    </AppContext>
  )
}

const InnerApp = observer(function InnerApp<S extends TLShape>(props: AppProps<S>): JSX.Element {
  const app = useApp<S>()
  return (
    <Renderer
      viewport={app.viewport}
      inputs={app.inputs}
      callbacks={app._events}
      brush={app.brush}
      hoveredShape={app.hoveredShape}
      selectedBounds={app.selectedBounds}
      selectedShapes={app.selectedShapes}
      shapes={app.shapesInViewport}
      showGrid={app.showGrid}
      showBounds={app.showBounds}
      showBoundsRotation={app.showBoundsRotation}
      showResizeHandles={app.showResizeHandles}
      showRotateHandle={app.showRotateHandle}
      showBoundsDetail={app.showBoundsDetail}
      showContextBar={app.showContextBar}
      {...props}
    />
  )
})
