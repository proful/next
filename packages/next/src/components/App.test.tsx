import * as React from 'react'
import { render, screen } from '@testing-library/react'
import { App } from './App'
import { TLApp, TLBoxShape } from '~nu-lib'

describe('app', () => {
  test('mounts component without crashing', () => {
    render(
      <App
        onMount={jest.fn()}
        onPersist={jest.fn()}
        model={mockDocument}
        shapeClasses={[TLBoxShape]}
      ></App>
    )
  })

  test('check the attributes', () => {
    render(
      <App
        onMount={jest.fn()}
        onPersist={jest.fn()}
        model={mockDocument}
        shapeClasses={[TLBoxShape]}
      ></App>
    )
    screen.logTestingPlaygroundURL()

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-ignore
    const app = tln as TLApp
    console.log(app.currentPageId)
  })
})

const mockDocument = {
  currentPageId: 'page1',
  selectedIds: ['box1'],
  pages: [
    {
      name: 'Page',
      id: 'page1',
      shapes: [
        {
          id: 'box1',
          type: 'box',
          parentId: 'page1',
          point: [100, 400],
          size: [100, 100],
        },
      ],
      bindings: [],
    },
  ],
}
