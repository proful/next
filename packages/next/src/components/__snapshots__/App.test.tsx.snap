// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app check the default attributes with only box shape 1`] = `
Object {
  "currentPageId": "page1",
  "pages": Array [
    Object {
      "bindings": Array [],
      "id": "page1",
      "name": "Page",
      "nonce": 1,
      "shapes": Array [
        Object {
          "children": undefined,
          "handles": undefined,
          "id": "box1",
          "isAspectRatioLocked": undefined,
          "isGenerated": undefined,
          "isGhost": undefined,
          "isHidden": undefined,
          "isLocked": undefined,
          "name": "Shape",
          "nonce": 0,
          "parentId": "page1",
          "point": Array [
            100,
            400,
          ],
          "rotation": undefined,
          "size": Array [
            100,
            100,
          ],
          "type": "box",
        },
      ],
    },
    Object {
      "bindings": Array [],
      "id": "page2",
      "name": "Page",
      "nonce": 1,
      "shapes": Array [
        Object {
          "children": undefined,
          "handles": undefined,
          "id": "box2",
          "isAspectRatioLocked": undefined,
          "isGenerated": undefined,
          "isGhost": undefined,
          "isHidden": undefined,
          "isLocked": undefined,
          "name": "Shape",
          "nonce": 0,
          "parentId": "page2",
          "point": Array [
            100,
            400,
          ],
          "rotation": undefined,
          "size": Array [
            100,
            100,
          ],
          "type": "box",
        },
      ],
    },
  ],
  "selectedIds": Array [
    "box1",
  ],
}
`;
